bchehraz/react-challenge-project-jan-2022
# Challenge 11 (Bonus): Login System

## Changes
1. (from challenge 1) Fixes order form input by changing `event.value` to `event.target.value`
2. (from challenge 4) Fixes authReducer action payload to properly login
3. (from challenge 5) Wires in the logout button in the `nav` component.
4. (from challenge 6) Wires in `PrivateRoute` route guard component
5. (from challenge 8) Wires in Add/Delete functionality
6. (from challenge 9) Creates new orderReducer and orderActions and is implemented into affected components.
7. Add `bcryptjs` and `jsonwebtoken` npm packages to the server for hashing passwords and generating json web tokens.
8. Add User model for mongoDB using mongoose
9. Add `/register` endpoint
10. Implements proper /login endpoint which now returns a valid JWT
11. Add `registerUser` redux action to `authActions.js`
12. Implements `registerForm` (basically a modified copy of login components) and adds the route to `appRouter`.
13. Adds link to `/register` page from `/login` page.
14. Add form validation validation to login form, which now displays error messages and a success message if login is successful.

## Purpose
- Link to issue: https://github.com/Shift3/react-challenge-project-jan-2022/issues/11
- Currently there is no way to create a user account, and plus the backend doesn't really log you in. We need a way to create an account, as well as actually verify the account exists before allowing you to log in.

## Approach
- The approach I took to solve this was based on previous experience setting up authentication in a node server. I ended up having both client side and server side form validation, to verify passwords match on register form, making sure the email is a valid email format, etc. 
- I then used `bcryptjs` to generate a hash, as well as check in `/login` if the password passed in matches the password set for the user they are attempting to log in as. 
- Additionally, I added in the package `jsonwebtoken` to the server to generate a jwt to be returned on a successful login-- however, the token is simply returned by the endpoint and is not set up to be passed via the `Authorization` header. In a real world scenario, this would be used on the client side as a means to communicate with the server via an `Authorization` header. With that, the server would need to read the token on each attempt to access a private route and only if the token is valid, allow the client access to such endpoints.
- The rest was simply creating/modifying the UI and wiring in logic with `/register`.

## Pre-Testing TODOs
- [ ] Install the latest server dependencies via docker 

## Testing Steps
1. Navigate to `/login`, click on the link at the bottom to navigate to the register page.
2. Test out the register form by typing in any combination of invalid emails mixed with passwords that don't match or no password at all. The form should provide feedback as to what is going wrong.
3. Now create a valid registration with a valid email and password+confirmation that match.
4. Now on the login form, type in any combination of wrong password or wrong email to see what error messages arise. 
6. Then type in the correct login information as specified when you registered. You should now gain access to the app!
7. As an optional, you can also test out registering an email that is already taken. Go to the `/register` route again. Type in an email that already exists and any matching passwords. There should be an error message now saying the email is already signed up.

## Learning
- I referenced a previous team project of mine which used MongoDB/mongoose. Although the backend of this project used GraphQL, the methods used for authenticating and creating a new user are the same. https://github.com/chingu-voyages/v25-bears-team-06
- Also, I ran into an issue installing the npm package `bcrypt` on docker, even though it worked fine in my npm environment. Apparently installing `bcryptjs` instead did the trick! Reference: https://stackoverflow.com/questions/29320201/error-installing-bcrypt-with-npm.

Closes https://github.com/Shift3/react-challenge-project-jan-2022/issues/11
